File: joss/comparisons/cat-6-contextual-runtime/08-async-sync/example-async-sync.R
# Error Class 8: Async/Sync Confusion
#
# R has no native async/await. All operations are synchronous.
# There is no type-level distinction between sync and async.
# External API calls require explicit HTTP libraries (httr2, curl).

library(tidyverse)

meds <- read_csv("fixtures/medications.csv")

# R does not have async DataFrame operations.
# To call an external API for each row, you must use map/lapply
# with blocking HTTP calls. There is no compile-time protection
# against mixing sync and async logic.

# Example: blocking API call per row (no async alternative)
# lookup_drug_interaction <- function(drug_name) {
#   response <- httr2::request("https://api.example.com/interactions") |>
#     httr2::req_url_query(drug = drug_name) |>
#     httr2::req_perform()
#   httr2::resp_body_json(response)$result
# }
#
# meds %>%
#   mutate(interaction = map_chr(medication_name, lookup_drug_interaction))
#
# No type checking on the return value. If the API returns NULL
# or a list instead of a string, R silently coerces or errors at runtime.

File: joss/comparisons/cat-6-contextual-runtime/08-async-sync/example-async-sync.py
"""
Error Class 8: Async/Sync Confusion

Python/pandas has no async-aware DataFrame operations. Mixing async
functions into DataFrame workflows requires manual await handling.
There is no type-level distinction between sync and async transforms.
"""
import pandas as pd
import asyncio

meds = pd.read_csv("fixtures/medications.csv")


async def lookup_drug_interaction(drug_name: str) -> str:
    """Simulates an async API call."""
    return f"No interactions found for {drug_name}"


# ── PROBLEM 8a: No async support in pandas ──────────────────────────────
# pandas .apply() does not support async functions. If you pass one,
# it returns a column of coroutine objects — not the actual results.
meds["interaction"] = meds["medication_name"].apply(lookup_drug_interaction)
print(meds["interaction"].iloc[0])
# Prints: <coroutine object lookup_drug_interaction at 0x...>
# No error, no warning — just wrong data silently stored.

# ── PROBLEM 8b: Manual async loop required ──────────────────────────────
# To actually run async functions, you must manually manage the event loop.
# This is error-prone and has no type-level guardrails.
async def process_meds():
    results = []
    for _, row in meds.iterrows():
        result = await lookup_drug_interaction(row["medication_name"])
        results.append(result)
    meds["interaction"] = results

asyncio.run(process_meds())
